src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,191 @@
+    pub connect_timeout: u64,
+}
+impl ConfigImpl for Config {}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "gsub".into()
+    }
+
+    async fn build(&self, alias: &str, raw_config: &ConnectorConfig) -> Result<Box<dyn Connector>> {
+        if let Some(raw) = &raw_config.config {
+            let config = Config::new(raw)?;
+            Ok(Box::new(GSub { config }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(alias.to_string()).into())
+        }
+    }
+}
+
+struct GSub {
+    config: Config,
+}
+
+struct GSubSource {
+    config: Config,
+    client: Option<SubscriberClient<InterceptedService<Channel, AuthInterceptor>>>,
+    ack_ids: DashMap<u64, String>,
+    ack_counter: AtomicU64,
+}
+
+impl GSubSource {
+    pub fn new(config: Config) -> Self {
+        GSubSource {
+            config,
+            client: None,
+            ack_ids: DashMap::new(),
+            ack_counter: AtomicU64::new(0),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Source for GSubSource {
+    async fn connect(&mut self, _ctx: &SourceContext, _attempt: &Attempt) -> Result<bool> {
+        let token = Token::new()?;
+
+        let tls_config = ClientTlsConfig::new()
+            .ca_certificate(Certificate::from_pem(googapis::CERTIFICATES))
+            .domain_name("pubsub.googleapis.com");
+
+        let channel = Channel::from_static("https://pubsub.googleapis.com")
+            .connect_timeout(Duration::from_nanos(self.config.connect_timeout))
+            .tls_config(tls_config)?
+            .connect()
+            .await;
+
+        let channel = dbg!(channel)?;
+
+        let client = SubscriberClient::with_interceptor(
+            channel,
+            AuthInterceptor {
+                token: Box::new(move || match token.header_value() {
+                    Ok(val) => Ok(val),
+                    Err(e) => {
+                        error!("Failed to get token for BigQuery: {}", e);
+
+                        Err(Status::unavailable(
+                            "Failed to retrieve authentication token.",
+                        ))
+                    }
+                }),
+            },
+        );
+
+        self.client = Some(client);
+
+        Ok(true)
+    }
+
+    async fn pull_data(&mut self, pull_id: &mut u64, ctx: &SourceContext) -> Result<SourceReply> {
+        let client = self
+            .client
+            .as_mut()
+            .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                // fixme use an error for GPubSub
+                "The client is not connected",
+            ))?;
+
+        let response = client
+            .pull(PullRequest {
+                subscription: "projects/wf-gcp-us-tremor-sbx/subscriptions/test-subscription-a"
+                    .to_string(),
+                // fixme config?
+                max_messages: 1,
```

Comment:
Hmmm... the signature of `pull_data` only allows this to be 1, i see. Does this have an obvious performance impact? Should we maybe think about returning a `Result<Vec<SourceReply>>`? cc @Licenser